### PR TITLE
Create ParticleInstructions

### PR DESCRIPTION
**Pull request by LLytho**

### CODE REVIEW

**Comment by LLytho** [comment]
Example code: 

```js
onEvent("ponder.tag", (event) => {
    event.createTag("particle_test", "minecraft:blaze_powder", "Test name", "Some test description!", [
        "minecraft:blaze_powder",
    ]);
});

onEvent("ponder.registry", (event) => {
    /**
     * Prints all particle names in the client.txt which can be used for `scene.particles.simple()`.
     */
    event.printParticleNames();

    event
        .create("particle_testing", ["minecraft:blaze_powder"])
        .tag("kubejs:particle_test")
        .scene("particles", "How to particle", (scene, util) => {
            scene.showStructure();
            scene.idle(10);

            const pos = [4, 1.5, 4];
            const start = [0, 1, 0];
            const end = [2, 2, 3];
            const tickLength = 20;
            const idleLength = tickLength * 3;

            /**
             * Using `scene.particles.` returns an object to set additional options
             *
             * .density(number) -> Sets the density of the particles.
             * .gravity(number) -> Sets the gravity of the particles.
             * .hasPhysics(true or false) -> Sets if the particles have physics.
             * .stoppedByCollision(true or false) -> Sets if the particles stop by collision.
             * .roll(number) -> Do a barrell roll!
             * .scale(number) -> Scale the particles between 0.01 and 4.
             * .lifetime(number) -> Set the lifetime of the particles.
             *
             * .motion(vec3) -> Sets the motion of the particles.
             * .speed(vec3) -> Sets the speed of the particles. Affects the motion if given.
             * .area(vec3) -> Sets the area of the particles.
             * .delta(vec3) -> Sets the delta of the particles. Same as the particle command.
             */
            scene.addKeyframe();
            scene.particles.simple(tickLength, "glow", pos);
            scene.particles.simple(tickLength, "glow", start).density(10).area(end);
            scene.idle(idleLength);

            scene.addKeyframe();
            scene.particles.simple(tickLength, "small_flame", pos);
            scene.particles.simple(tickLength, "small_flame", start).density(10).motion([0, 0, -0.1]).area(end);
            scene.idle(idleLength);

            scene.addKeyframe();
            scene.particles.item(tickLength, "minecraft:diamond_block", pos).motion([-0.09, 0.3, 0]).density(8);
            scene.particles.item(tickLength, "minecraft:diamond_block", start).area(end);
            scene.idle(idleLength);

            scene.addKeyframe();
            scene.particles.block(tickLength, "minecraft:diamond_block", pos);
            scene.particles.block(tickLength, "minecraft:diamond_block", start).density(4).area(end);
            scene.idle(idleLength);

            scene.addKeyframe();
            scene.particles.dust(tickLength, "#00FFF0", start).density(5).motion([0, 0, -0.1]).area(end).roll(10);
            scene.idle(idleLength);

            scene.addKeyframe();
            scene.particles
                .dust(tickLength, "#FF0000", "#0000FF", start)
                .density(2)
                .scale(2.1)
                .motion([0, 0, -0.1])
                .area(end)
                .roll(3);
            scene.idle(idleLength);
        });
});

```

File: build.gradle [hunk 1]
```diff
@@ -136,10 +136,10 @@ jar {
     manifest {
         attributes([
                 "Specification-Title"     : mod_id,
-                "Specification-Vendor"    : mod_authors,
+                "Specification-Vendor"    : project.version,
```
Comment (rlnt) [inline build.gradle]: `vendor` is something like "Herausgeber", why is the version here?

File: build.gradle [hunk 2]
```diff
@@ -136,10 +136,10 @@ jar {
...
                 "Specification-Version"   : "1",
                 "Implementation-Title"    : project.name,
-                "Implementation-Version"  : mod_version,
+                "Implementation-Version"  : project.version,
```
Comment (rlnt) [inline build.gradle:142]: this should not make any difference

File: build.gradle (hunk 1 above)
Comment (LLytho) [reply build.gradle]: I removed the wrong one, fuck

File: src/main/java/com/almostreliable/ponderjs/mixin/SceneBuilderMixin.java [hunk 3]
```diff
@@ -96,6 +101,14 @@ public abstract class SceneBuilderMixin {
         return special;
     }
 
+    @RemapForJS(("getParticles"))
```
Comment (rlnt) [inline src/main/java/com/almostreliable/ponderjs/mixin/SceneBuilderMixin.java]: remove these unnecessary double parenthesis

File: src/main/java/com/almostreliable/ponderjs/PonderRegistryEventJS.java [hunk 4]
```diff
@@ -16,4 +19,17 @@ public PonderBuilderJS create(String name, IngredientJS ingredient) {
...
         return new PonderBuilderJS(name, ingredient.getVanillaItems());
     }
+
+    public void printParticleNames() {
+        StringBuilder sb = new StringBuilder();
+        sb.append("\n").append("### Particles ###").append("\n");
+        //noinspection ConstantConditions
+        ForgeRegistries.PARTICLE_TYPES.getValues()
+                .stream()
+                .filter(p -> p instanceof SimpleParticleType)
```
Comment (rlnt) [inline src/main/java/com/almostreliable/ponderjs/PonderRegistryEventJS.java]: replace this filter with a method reference `SimpleParticleType.class::isInstance`

File: src/main/java/com/almostreliable/ponderjs/mixin/PonderWorldAccessor.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+import net.minecraft.client.particle.Particle;
+import net.minecraft.core.particles.ParticleOptions;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.gen.Invoker;
+
+@Mixin(PonderWorld.class)
+public interface PonderWorldAccessor {
+
+    @Invoker(value = "makeParticle", remap = false)
+    <T extends ParticleOptions>Particle ponderjs$makeParticle(T data, double x, double y, double z, double mx, double my, double mz);
```
Comment (rlnt) [inline src/main/java/com/almostreliable/ponderjs/mixin/PonderWorldAccessor.java]: whitespace after the generic is missing

File: src/main/java/com/almostreliable/ponderjs/particles/ParticleDataBuilder.java [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+    @SuppressWarnings("unchecked")
+    protected O getSelf() {
+        return (O) this;
+    }
+
+    public static class Static extends ParticleDataBuilder<Static, ParticleOptions> {
+        private final ParticleOptions type;
+
+        public Static(ParticleOptions type) {
+            super();
```
Comment (rlnt) [inline src/main/java/com/almostreliable/ponderjs/particles/ParticleDataBuilder.java]: empty super call is unnecessary

File: src/main/java/com/almostreliable/ponderjs/particles/ParticleInstructions.java [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+                    scene.getWorld().addParticle(particle);
+                }
+            }
+        }
+
+        private void applyParticleData(Particle particle) {
+            if (particle instanceof ParticleAccessor accessor) {
+                if (builder.color != null) {
+                    long argb = builder.color.getArgbKJS();
+                    float a = (float) (argb >> 24 & 255) / 255.0F;
```
Comment (rlnt) [inline src/main/java/com/almostreliable/ponderjs/particles/ParticleInstructions.java]: casting to float is unnecessary if you divide by a float

File: src/main/java/com/almostreliable/ponderjs/particles/ParticleInstructions.java [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+        this.scene = scene;
+    }
+
+    public ParticleDataBuilder<?, ?> simple(int ticks, ParticleType<?> type, Vec3 pos) {
+        if (type instanceof SimpleParticleType simple) {
+            return create(ticks, pos, new ParticleDataBuilder.Static(simple));
+        }
+
+        throw new IllegalArgumentException(
+                "Particle type " + (type == null ? "INVALID" : Registry.PARTICLE_TYPE.getKey(type)) +
```
Comment (rlnt) [inline src/main/java/com/almostreliable/ponderjs/particles/ParticleInstructions.java]: is the usage of a deprecated registry intended here?

File: src/main/java/com/almostreliable/ponderjs/particles/ParticleTransformation.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+                Object first = Context.jsToJava(list.get(0), Vec3.class);
+                Object second = Context.jsToJava(list.get(1), Vec3.class);
+
+                if (first instanceof Vec3 pos && second instanceof Vec3 motion) {
+                    return new Data(pos, motion);
+                }
+            }
+
+            IllegalArgumentException e = new IllegalArgumentException(
+                    "Invalid format for particle transformation data. Please use [position, motion] or [ [x, y, z], [mx, my, mz]]");
```
Comment (rlnt) [inline src/main/java/com/almostreliable/ponderjs/particles/ParticleTransformation.java]: remove the whitespaces after the first opening bracket or add one between the closing brackets for consistency

**Review by rlnt — CHANGES_REQUESTED** [review]

**Review by rlnt — APPROVED** [review]

**Pull request merged**